ports/esp32s2/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,199 @@
+                                                    uint32_t frequency,
+                                                    bool variable_frequency) {
+    // Calculate duty cycle
+    uint32_t duty_bits = 0;
+    uint32_t interval = LEDC_APB_CLK_HZ/frequency;
+    for (size_t i = 0; i < 32; i++) {
+        if(!(interval >> i)) {
+            duty_bits = i - 1;
+            break;
+        }
+    }
+    if (duty_bits < 1) {
+        mp_raise_ValueError(translate("Invalid frequency"));
+    } else if (duty_bits >= LEDC_TIMER_14_BIT) {
+        duty_bits = LEDC_TIMER_13_BIT;
+    }
+
+    // Find a viable timer
+    size_t timer_index = INDEX_EMPTY;
+    size_t channel_index = INDEX_EMPTY;
+    for (size_t i = 0; i < LEDC_TIMER_MAX; i++) {
+        if ((reserved_timer_freq[i] == frequency) && !variable_frequency) {
+            //prioritize matched frequencies so we don't needlessly take slots
+            timer_index = i;
+            break;
+        } else if (reserved_timer_freq[i] == 0) {
+            timer_index = i;
+            break;
+        }
+    }
+    if (timer_index == INDEX_EMPTY) {
+        // Running out of timers isn't pin related on ESP32S2 so we can't re-use error messages
+        mp_raise_ValueError(translate("No more timers available"));
+    }
+
+    // Find a viable channel
+    for (size_t i = 0; i < LEDC_CHANNEL_MAX; i++) {
+        if (reserved_channels[i] == INDEX_EMPTY) {
+            channel_index = i;
+            break;
+        }
+    }
+    if (channel_index == INDEX_EMPTY) {
+        mp_raise_ValueError(translate("No more channels available"));
+    }
+
+    // Run configuration
+    self->tim_handle.timer_num = timer_index;
+    self->tim_handle.duty_resolution = duty_bits;
+    self->tim_handle.freq_hz = frequency;
+    self->tim_handle.speed_mode = LEDC_LOW_SPEED_MODE;
+    self->tim_handle.clk_cfg = LEDC_AUTO_CLK;
+
+    if (ledc_timer_config(&(self->tim_handle)) != ESP_OK) {
+        mp_raise_ValueError(translate("Could not initialize timer"));
+    }
+
+    self->chan_handle.channel = channel_index;
+    self->chan_handle.duty = duty >> (16 - duty_bits);
+    self->chan_handle.gpio_num = pin->number;
+    self->chan_handle.speed_mode = LEDC_LOW_SPEED_MODE; // Only LS is allowed on ESP32-S2
+    self->chan_handle.hpoint = 0;
+    self->chan_handle.timer_sel = timer_index;
+
+    if (ledc_channel_config(&(self->chan_handle))) {
+        mp_raise_ValueError(translate("Could not initialize channel"));
+    }
+
+    // Make reservations
+    reserved_timer_freq[timer_index] = frequency;
+    reserved_channels[channel_index] = timer_index;
+
+    self->variable_frequency = variable_frequency;
+    self->pin_number = pin->number;
+    self->deinited = false;
+    self->duty_resolution = duty_bits;
+    claim_pin(pin);
+
+    // Set initial duty
+    ledc_set_duty(LEDC_LOW_SPEED_MODE, self->chan_handle.channel, duty >> (16 - duty_bits));
+    ledc_update_duty(LEDC_LOW_SPEED_MODE, self->chan_handle.channel);
```

Comment:
Just call set_duty_cycle to remove duplication.

Reply:
Do you mean `ledc_hal_set_duty_cycle`? I would need to create an analog of the internal `ledc_hal_context_t` pointer to use it, and it's not included a part of ESP's API Reference as an internal function. Unless you meant make my own, which I'll assume unless you say otherwise.

Reply:
I meant common_hal_pulseio_pwmout_set_duty_cycle. No need for a separate static function.

Reply:
Wow that went right over my head. Yes, good idea.